vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SqlConnectionFactoryBase.java

```diff
@@ -75,35 +80,79 @@ protected SqlConnectionFactoryBase(VertxInternal vertx, SqlConnectOptions option
     this.netClient = vertx.createNetClient(netClientOptions, clientCloseFuture);
   }
 
+  private int advanceNextHost() {
+    return lastLiveHost.getAndIncrement() % socketAddresses.size();
+  }
+
+  private static class AddressesIterator implements Iterator<SocketAddress> {
+    private final int firstAddress;
+    private int currentAddress;
+    private final List<SocketAddress> socketAddresses;
+
+    AddressesIterator(int firstAddress, List<SocketAddress> socketAddresses) {
+      this.firstAddress = firstAddress;
+      this.currentAddress = firstAddress;
+      this.socketAddresses = socketAddresses;
+    }
+
+    private int nextAddress() {
+      return currentAddress == socketAddresses.size() - 1 ? 0 : currentAddress + 1;
+    }
+
+    /**
+     * Not a proper Iterator!
+     * if socketAddresses contains only one element this will return false on first call
+     * {@link SqlConnectionFactoryBase#doConnectWithRetry} always calls next() on first attempt without checking hasNext()
+     * making this implementation sufficient
+     */
+    @Override
+    public boolean hasNext() {
+      return firstAddress != nextAddress();
+    }
+
+    @Override
+    public SocketAddress next() {
+      SocketAddress socketAddress = socketAddresses.get(currentAddress);
+      currentAddress = nextAddress();
+      return socketAddress;
+    }
+  }
+
+  private Iterator<SocketAddress> connectionOrder() {
+    return new AddressesIterator(advanceNextHost(), socketAddresses);
+  }
+
   @Override
   public void connect(Promise<Connection> promise) {
     PromiseInternal<Connection> promiseInternal = (PromiseInternal<Connection>) promise;
     ContextInternal context = promiseInternal.context();
-    context.emit(promise, p -> doConnectWithRetry(promiseInternal, reconnectAttempts));
+    context.emit(promise, p -> doConnectWithRetry(promiseInternal, reconnectAttempts, connectionOrder()));
   }
 
   @Override
   public void close(Promise<Void> promise) {
     clientCloseFuture.close(promise);
   }
 
-  private void doConnectWithRetry(PromiseInternal<Connection> promise, int remainingAttempts) {
+  private void doConnectWithRetry(PromiseInternal<Connection> promise, int remainingAttempts, Iterator<SocketAddress> addresses) {
     ContextInternal context = promise.context();
-    Promise<Connection> promise0 = context.promise();
+    PromiseInternal<Connection> promise0 = context.promise();
     promise0.future().onComplete(ar -> {
       if (ar.succeeded()) {
         promise.complete(ar.result());
       } else {
-        if (remainingAttempts >= 0) {
+        if (addresses.hasNext()) {
+          doConnectWithRetry(promise, remainingAttempts, addresses);
```

Comment:
This is the gist of this PR. I decided to keep this logic in current method to not to create additional promises.